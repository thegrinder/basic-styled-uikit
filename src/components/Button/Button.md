```js
<Button>Default button</Button>
```