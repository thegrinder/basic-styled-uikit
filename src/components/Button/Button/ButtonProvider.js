import { createProvider } from '../../../helpers';
import { buttonBaseTheme, buttonDarkTheme } from './buttonTheme';

const ButtonProvider = createProvider(
  'button',
  buttonBaseTheme,
  buttonDarkTheme
);

export default ButtonProvider;
