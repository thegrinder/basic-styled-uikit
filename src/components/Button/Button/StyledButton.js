import styled, { css } from 'styled-components';
import { string } from 'prop-types';
import { rem } from '../../Typography';
import {
  getBtnTypeStyle,
  getBtnSizingStyle,
  getBtnCommonStyle,
} from './buttonSelectors';

const propTypes = {
  btntype: string.isRequired,
  sizing: string.isRequired,
};

const StyledButton = styled.button`
  border: none;
  overflow: visible;
  position: relative;
  display: inline-block;
  vertical-align: middle;
  text-decoration: none;
  transition: 0.1s ease-in-out;
  transition-property: color, background-color, border-color;
  cursor: pointer;
  outline: none;
  border-width: 1px;
  border-style: solid;
  border-color: transparent;
  background-color: transparent;
  border-radius: ${rem(6)};
  &:disabled {
    opacity: 0.6;
    cursor: default;
  }
  ${({ theme, btntype, sizing }) => css`
    ${getBtnCommonStyle(theme)}
    ${getBtnSizingStyle(theme, sizing)}
    ${getBtnTypeStyle(theme, btntype, 'normal')}
    &:hover,
    &:focus {
      ${getBtnTypeStyle(theme, btntype, 'hover')}
    }
    &:active {
      ${getBtnTypeStyle(theme, btntype, 'active')}
    }
    &:disabled {
      ${getBtnTypeStyle(theme, btntype, 'normal')}
    }
  `}
`;

StyledButton.propTypes = propTypes;

export default StyledButton;
