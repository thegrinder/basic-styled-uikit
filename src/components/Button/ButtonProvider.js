import { createProvider } from '../../helpers';

const minimalTheme = {
  btnTypes: {
    default: {},
  },
  sizings: {
    m: {},
  },
};

const ButtonProvider = createProvider('button', minimalTheme);

export default ButtonProvider;
