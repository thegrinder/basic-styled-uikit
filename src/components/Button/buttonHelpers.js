import { getStyle } from '../../helpers/utils';

const getBtnType = ({ theme, btnType }, state) => theme.uiKit.button.btnTypes[btnType][state];
const getBtnTypeStyle = state => getStyle(getBtnType, state);
export const btnNormalStyle = getBtnTypeStyle('normal');
export const btnHoverStyle = getBtnTypeStyle('hover');
export const btnActiveStyle = getBtnTypeStyle('active');

const getBtnSize = ({ theme, btnSize }) => theme.uiKit.button.btnSizes[btnSize];
export const btnSizeStyle = getStyle(getBtnSize);
