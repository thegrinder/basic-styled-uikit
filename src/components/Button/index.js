export { default as ButtonProvider } from './ButtonProvider';
export { default as Button } from './Button';
