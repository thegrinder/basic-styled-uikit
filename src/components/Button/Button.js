import React from 'react';
import { node, bool, string, oneOfType, oneOf, func } from 'prop-types';
import styled from 'styled-components';
import StyledButton from './StyledButton';

const ContentWrapper = styled.span`
  display: flex;
  align-items: center;
  opacity: ${({ submitting }) => (submitting ? '0' : '1')};
`;

ContentWrapper.propTypes = {
  submitting: bool.isRequired,
};

const SpinnerWrapper = styled.span`
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const propTypes = {
  /** submitting flag that shows a component passed to renderSpinner prop */
  submitting: bool,
  /** one of: default, primary, danger */
  btnType: string,
  /** one of: s, m, l */
  sizing: string,
  /** spinner component shown when submitting flag is true */
  renderSpinner: node,
  /** rendered html tag or custom router link component */
  as: oneOfType([oneOf(['a', 'button']), func]),
  children: node,
};

const Button = ({
  btnType: btntype = 'default',
  sizing = 'm',
  submitting = false,
  renderSpinner,
  children,
  as = 'button',
  ...rest
}) => (
  <StyledButton as={as} sizing={sizing} btntype={btntype} {...rest}>
    {submitting && <SpinnerWrapper>{renderSpinner}</SpinnerWrapper>}
    <ContentWrapper submitting={submitting}>{children}</ContentWrapper>
  </StyledButton>
);

Button.propTypes = propTypes;

export default Button;
