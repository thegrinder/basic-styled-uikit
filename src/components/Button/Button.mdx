---
name: Button
menu: Components
---

import { Playground, PropsTable } from 'docz';
import Button from './Button';
import { Spacer, Spinner, Icon } from '../../../docz';

## Button

<PropsTable of={Button} />


### Button types

<Playground>
  <Button className="mb">default</Button>
  <Spacer />
  <Button btnType="primary">primary</Button>
  <Spacer />
  <Button btnType="danger">danger</Button>
</Playground>


### Button sizes

<Playground>
  <Button btnSize="small">small</Button>
  <Spacer />
  <Button>default</Button>
  <Spacer />
  <Button btnSize="large">large</Button>
</Playground>


### Disabled button

<Playground>
  <Button btnType="primary" disabled={true}>disabled</Button>
</Playground>


### Disabled Button with a spinner

<Playground>
  <Button
    btnType="primary"
    disabled={true}
    submitting={true}
    renderSpinner={<Spinner />}
  >
    submitting
  </Button>
</Playground>


### Button with icon

<Playground>
  <Button left={<Icon />}>left icon</Button>
  <Spacer />
  <Button right={<Icon />}>right icon</Button>
</Playground>
