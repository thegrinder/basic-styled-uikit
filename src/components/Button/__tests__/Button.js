import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render, cleanup } from 'react-testing-library';
import 'jest-dom/extend-expect';

import theme from '../../../theme/theme';
import Button from '../Button';


const children = <span>children</span>;

const renderComponent = (props = {}) => render(
  <ThemeProvider theme={theme}>
    <Button {...props}>{children}</Button>
  </ThemeProvider>,
);

describe('<Button />', () => {
  afterEach(cleanup);

  it('should render correctly with default props and children', () => {
    const { container: { firstChild }, getByText } = renderComponent();
    const childrenElement = getByText('children');
    expect(firstChild).toBeDefined();
    expect(firstChild).toContainElement(childrenElement);
    expect(firstChild).toMatchSnapshot();
  });

  it('should render correctly with custom props', () => {
    const left = <span>left</span>;
    const right = <span>right</span>;
    const spinner = <span data-testid="spinner" />;
    const { container: { firstChild }, getByText, getByTestId } = renderComponent({
      btnType: 'danger',
      sizing: 's',
      left,
      right,
      submitting: true,
      renderSpinner: spinner,
    });

    const leftElement = getByText('left');
    const rightElement = getByText('right');
    const spinnerElement = getByTestId('spinner');
    const childrenElement = getByText('children');

    expect(firstChild).toBeDefined();
    expect(firstChild).toContainElement(leftElement);
    expect(firstChild).toContainElement(rightElement);
    expect(firstChild).toContainElement(spinnerElement);
    expect(firstChild).toContainElement(childrenElement);
    expect(firstChild).toMatchSnapshot();
  });

  it('should render <a> tag by default', () => {
    const { container: { firstChild } } = renderComponent({ as: 'a' });
    expect(firstChild.tagName).toEqual('A');
    expect(firstChild).toMatchSnapshot();
  });

  it('should render render <button> tag', () => {
    const { container: { firstChild } } = renderComponent();
    expect(firstChild.tagName).toEqual('BUTTON');
    expect(firstChild).toMatchSnapshot();
  });
});
