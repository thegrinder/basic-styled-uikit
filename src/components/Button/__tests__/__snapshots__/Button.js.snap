// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render correctly with custom props 1`] = `
<button
  class="StyledButton-k0ad56-0 cpgfNg"
>
  <span
    class="SpinnerWrapper-sc-1ln0y9p-0 csgNUt"
  >
    <span
      data-testid="spinner"
    />
  </span>
  <span
    class="ContentWrapper-juvto2-0 haOASu"
  >
    <span>
      children
    </span>
  </span>
</button>
`;

exports[`<Button /> should render correctly with default props and children 1`] = `
<button
  class="StyledButton-k0ad56-0 ixFtWQ"
>
  <span
    class="ContentWrapper-juvto2-0 kMRFXB"
  >
    <span>
      children
    </span>
  </span>
</button>
`;
